File: java-spring/architecture-shell-cqrs/src/main/java/com/architecture/shell/cqrs/RequestHandler.java
package com.architecture.shell.cqrs;

/**
 * Base handler interface for processing requests.
 * All command and query handlers extend this interface.
 *
 * Requirements:
 * - FR-003: Each request type maps to exactly ONE handler (enforced at DI registration)
 * - FR-005: Handlers MUST support cancellation/timeout semantics
 * - BR-007: Handlers use Result<T> for business errors, throw exceptions for infrastructure errors
 *
 * Handler Lifecycle:
 * 1. Mediator resolves handler for request type
 * 2. Pipeline behaviors wrap handler execution
 * 3. Handler processes request
 * 4. Result returned or exception propagated
 *
 * Error Handling:
 * - Business errors (validation failures, domain rules): Return Result.failure()
 * - Infrastructure errors (DB down, network timeout): Throw exception
 *
 * DDD Layer: Application Layer
 *
 * @param <TRequest>  The request type
 * @param <TResponse> The response type
 * @see CommandHandler
 * @see QueryHandler
 */
@FunctionalInterface
public interface RequestHandler<TRequest extends BaseRequest, TResponse> {

    /**
     * Handles the request and returns a response.
     *
     * @param request The request to process
     * @return The response (may be Result<T> for business error handling)
     * @throws RuntimeException for infrastructure errors (DB connectivity, network failures)
     */
    TResponse handle(TRequest request);
}

File: golang/architecture-shell-cqrs/request_handler.go
package cqrs

import "context"

// RequestHandler is the base handler interface for processing requests.
// Each request type maps to exactly one handler (enforced at mediator construction).
// Handlers MUST respect context cancellation and terminate early when requested.
// Handlers MUST use Result[T] for business errors, return error for infrastructure errors.
//
// Error Handling Pattern:
//   - Business errors: return Result[T].Failure(), nil
//   - Infrastructure errors: return zero value, error
type RequestHandler[TRequest BaseRequest, TResponse any] interface {
	// Handle processes the request.
	// ctx: Context for cancellation and deadline control
	// request: The request to process
	// Returns: (response, nil) on success, (zero, error) on infrastructure failure
	Handle(ctx context.Context, request TRequest) (TResponse, error)
}
